agent/agent.go

```diff
@@ -1068,9 +1067,34 @@ LOAD:
 	return nil
 }
 
+// RegisterEndpoint registers a handler for the consul RPC server
+// under a unique name while making it accessible under the provided
+// name. This allows overwriting handlers for the golang net/rpc
+// service which does not allow this.
+func (a *Agent) RegisterEndpoint(name string, handler interface{}) error {
+	srv, ok := a.delegate.(*consul.Server)
+	if !ok {
+		panic("agent must be a server")
+	}
+	realname := fmt.Sprintf("%s-%d", name, time.Now().UnixNano())
+	a.endpointsLock.Lock()
+	a.endpoints[name] = realname
+	a.endpointsLock.Unlock()
+	return srv.RegisterEndpoint(realname, handler)
+}
+
 // RPC is used to make an RPC call to the Consul servers
 // This allows the agent to implement the Consul.Interface
 func (a *Agent) RPC(method string, args interface{}, reply interface{}) error {
+	a.endpointsLock.Lock()
+	// fast path: only translate if there are overrides
+	if len(a.endpoints) > 0 {
+		p := strings.SplitN(method, ".", 2)
+		if e := a.endpoints[p[0]]; e != "" {
+			method = e + "." + p[1]
```

Comment:
Since `a.endpoints` is empty for normal operations this isn't relevant IMO. `a.endpoints` is only populated through `registerEndpoint` which is only used for testing. Also, the naming scheme of the RPC server dictates a two parts: the object and the method on the object.